docs/specification/auth/oauth2.md

```diff
@@ -0,0 +1,252 @@
+
+## Capabilities
+
+Clients supporting OAuth 2.0 **MUST** declare it during initialization:
+
+```json
+{
+  "capabilities": {
+    "auth": {
+      "oauth2": true
+    }
+  }
+}
+```
+
+Servers supporting OAuth 2.0 **MUST** include their capabilities:
+
+```json
+{
+  "capabilities": {
+    "auth": {
+      "oauth2": {
+        "authorize": true,
+        "token": true,
+        "revoke": true
+      }
+    }
+  }
+}
+```
+
+## Flows
+### Initialization
+During initialization, if the client and server both support the `oauth2` capability, the client **SHOULD** include an access token in all subsequent requests. If the client does not have an access token, the client **SHOULD** obtain one.
+
+### Obtaining an Access Token
+
+#### Authorization Grant Flow
+To obtain an access token, clients **MUST** send an `auth/oauth2/authorize` request.
+
+**Request:**
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "auth/oauth2/authorize",
+  "params": {
+    "response_type": "code", // REQUIRED
+    "client_id": "client_id", // REQUIRED
+    "redirect_uri": "redirect_uri", // OPTIONAL
+    "scope": "scope", // OPTIONAL
+    "state": "state", // RECOMMENDED
+  }
+}
+```
+**Response:**
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "authorization_url": "<scheme>://<domain>/oauth2"
+  }
+}
+```
+- The client **MUST** redirect the user to the authorization URL.
+- The authorization URL **SHOULD** prompt the user to configure any credentials as needed. Once complete, the server provides an authorization grant in the form of a code.
+- The client **MUST** implement a way to receive the authorization grant. This can be through a callback such as the redirect URI, providing some interface for the user to provide it, etc.
+
+#### Access Token Exchange Flow
+
+After receiving the authorization code, the client **SHOULD** exchange it for tokens:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "auth/oauth2/token",
+  "params": {
+    "grant_type": "authorization_code",
+    "code": "code",
+    "state": "state"
+  }
+}
+```
+
+The server responds with tokens:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 2,
+  "result": {
+    "access_token": "access_token", // REQUIRED
+    "token_type": "token_type", // REQUIRED
+    "expires_in": 3600, // RECOMMENDED
+    "scope": "scope", // OPTIONAL if idential to client-requested scope, otherwise REQUIRED
+    "state": "state" // REQUIRED if the "state" parameter was present in the client authorization request.
+    // OPTIONAL additional parameters
+  }
+}
+```
+- The client **MUST** securely store the access token, and a refresh token if one is included.
+- The server **MUST** implement rate-limiting to prevent abuse.
+
+### Utilizing an Access Token
+
+Once a client has obtained an access token, it **SHOULD** include it in all requests in the parameters, including the initalization request.
```

Comment:
Is there a technical reason why we need to include it in all requests? (Plausibly yes, just want to check.)

Reply:
I envisioned that the server would present different behaviors (more capabilities, different tool functionality, etc.) if the client was authenticated. For ease of implementation, I wrote that the client should (but is not required to) include it in all requests.